Clarify encoding of colon between scheme and type (#361)
package-url/purl-spec: PURL-SPECIFICATION.rst

- **scheme**:

  - The ``scheme`` is a constant with the value "pkg".

>   - The ``scheme`` and ``type`` MUST be separated by a colon ':'.

Author: ```suggestion
  - The ``scheme`` MUST be followed by a unencoded colon ':'.
```

Reviewer: Good suggestion -- change made.

>   - ``purl`` parsers MUST accept URLs in which the ``scheme`` and colon ':' are

Author: ```suggestion
  - ``purl`` parsers MUST accept URLs where the ``scheme`` and colon ':' are
```

    followed by one or more slash '/' characters, such as 'pkg://', and MUST

>     ignore -- i.e., normalize by removing -- all such '/' characters.

Author: ```suggestion
    ignore and remove all such '/' characters.
```

Reviewer: Good clarification, change made, and I'm making a corresponding change to related language in the FAQ.

Character encoding
~~~~~~~~~~~~~~~~~~

> For clarity and simplicity, a ``purl`` is always an ASCII string. To ensure that

Author: We need to refactor this section separately. @mprpic volunteered to make this PR.
Let's remove all changes to this Character encoding section from this PR.

Reviewer: @pombredanne I've backed out all of my prior changes to the "Character encoding" section as requested.  In light of @mprpic 's PR #389, I've also deleted a reference to the "Character encoding" section in the FAQ and made a few other adjustments.

For clarity and simplicity a ``purl`` is always an ASCII string. To ensure that
there is no ambiguity when parsing a ``purl``, separator characters and non-ASCII

> characters MUST be UTF-encoded and then percent-encoded as defined at::

Author: ```suggestion
characters must be UTF-encoded and then percent-encoded as defined at::
```

- the '#', '?', '@' and ':' characters must NOT be encoded when used as

>   separators. They may need to be encoded elsewhere

Reviewer 2: To fully address the confusion of #39, I would probably also change the last sentence here:

```suggestion
  separators. Some of them need to be encoded elsewhere as specified in the rules below.
```

> - The colon ':' separator between ``scheme`` and ``type`` MUST NOT be encoded.

Author: I guess that following @JimFuller-RedHat suggestion in today's triage call https://github.com/package-url/purl-spec/discussions/360 ... this could reworded as:
```suggestion
- The colon ':' separator between ``scheme`` and ``type`` MUST be used as-is, unencoded.
```

@JimFuller-RedHat ?

Reviewer 3: Hehe, I was (badly) trying say that specs typically define grammar, schema and behavior (ex. processing rules) ... in this case I think how you had it was fine. It is easier to define a 'thing' precisely which has the side effect of defining all the things it cannot be ;)

From grammar pov we could define (in EBNF) as a literal colon ... any parser built with such an EBNF would bork on encoded input ... we might suggest defining pre processing rules to normalise odd input into canonical purl. or define a looser grammar allowing encoding. 

I would propose:

_"The``scheme`` and ``type`` MUST be separated by a colon ':' "_

Author: @JimFuller-RedHat 
> I would propose:
> 
> _"The`scheme` and `type` MUST be separated by a colon ':' "_

clean and neat.

Yet someone would likely immediately bet a nickel that someone else will open an issue asking whether the colon can or must or must not be encoded. ;)

Here is the way out:
We can keep these negative assertions as comments in a new FAQ document. I think that we can create a leaner spec this way.

I think it makes sense for the spec to avoid specifying all the things that it does not support as you rightly suggest, because there is an infinity of these. And we can push "negative" support questions in that FAQ.

So here:

- In the spec:
  - `The scheme and type MUST be separated by a colon ':'`

- In the FAQ:
  - Q: Is the colon between scheme and type encoded? Or can it be encoded? If yes, how?
  - A: The spec has no mention of encoding there, so the colon should be used as-is, never encoded and never requiring any decoding since this is not encoded. It is a parsing error if the colon does not come directly after `pkg`; Lenient parsing tools are welcome to recover from this error to help process and sanitize damaged purls, but that's not a requirement, and not part of the spec.

That approach can become the standard. The spec will be leaner and crisper and a the FAQ comprehensive to address the questions that _will_ come up for sure.

Reviewer 3: cool ! fantastic suggestion - brevity and conciseness enables implementators to focus ... a good example is the json spec - https://ecma-international.org/wp-content/uploads/ECMA-404_2nd_edition_december_2017.pdf , which only contains normative text ;)

  For example, in the PURL snippet ``pkg:npm`` the colon ':' MUST NOT be encoded,
  and the PURL snippet ``pkg%3Anpm`` is invalid.

> 

Reviewer: @gernot-h   @pombredanne   Consider adding at the top of the file, perhaps as a new one-line paragraph following the current first paragraph, something along the lines of the following:

> This specification uses RFC 2119 (https://datatracker.ietf.org/doc/html/rfc2119), as clarified in RFC 8174 (https://datatracker.ietf.org/doc/html/rfc8174), for the interpretation of certain terms, e.g., ``MUST NOT``.

Or perhaps a slight modification to the example provided by RFC 2119:

> The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119), as clarified in [RFC 8174](https://datatracker.ietf.org/doc/html/rfc8174).

(Note that the core spec currently contains a great deal of language that will need to be modified to implement RFC 2119/8174.)
